Build release artifacts for Alpine

Alpine uses musl. Rust-analyzer in VSCode started shipping a glibc linked binary recently.

NixOS had the same issue (#4542), but the fix only applied to NixOS, apparently.

This is pretty disappointing given that Rust 1.44 finally added support for proc macros in Alpine... but now rust-analyzer is broken.

I can't build anything with `-crt-static`:

```
  = note: "gcc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-Wl,--eh-frame-hdr" "-L" "/home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib" "/home/grayshade/hello/target/x86_64-unknown-linux-musl/release/deps/hello-a7c96442787521aa.hello.37986wrc-cgu.0.rcgu.o" "/home/grayshade/hello/target/x86_64-unknown-linux-musl/release/deps/hello-a7c96442787521aa.hello.37986wrc-cgu.1.rcgu.o" "/home/grayshade/hello/target/x86_64-unknown-linux-musl/release/deps/hello-a7c96442787521aa.hello.37986wrc-cgu.2.rcgu.o" "-o" "/home/grayshade/hello/target/x86_64-unknown-linux-musl/release/deps/hello-a7c96442787521aa" "/home/grayshade/hello/target/x86_64-unknown-linux-musl/release/deps/hello-a7c96442787521aa.541zeg1d0c7iasv1.rcgu.o" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-Wl,-O1" "-nodefaultlibs" "-L" "/home/grayshade/hello/target/x86_64-unknown-linux-musl/release/deps" "-L" "/home/grayshade/hello/target/release/deps" "-L" "/home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib" "-Wl,--start-group" "-Wl,-Bstatic" "/home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libstd-60e6d7e96d924282.rlib" "/home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libpanic_unwind-44b8a8ebcf9a9b46.rlib" "/home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libminiz_oxide-72a65827cdbd0fcb.rlib" "/home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libadler-857ab2ee733e2f82.rlib" "/home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libobject-76e3877305158d04.rlib" "/home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libaddr2line-8eb04fa574b28244.rlib" "/home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libgimli-c992d9f8b5e83846.rlib" "/home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/librustc_demangle-2f387d801141c407.rlib" "/home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libhashbrown-c3b261c6f6ba7e9a.rlib" "/home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/librustc_std_workspace_alloc-bd54d25606272510.rlib" "/tmp/rustcJ1tnuW/libunwind-5224c3e089786f59.rlib" "/home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libcfg_if-882a76520c66e911.rlib" "/tmp/rustcJ1tnuW/liblibc-a6df7fa84d406b53.rlib" "/home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/liballoc-85d3711e9a39f31b.rlib" "/home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/librustc_std_workspace_core-e7a65e8224e3fd05.rlib" "/home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libcore-162d017b99a78519.rlib" "-Wl,--end-group" "/home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libcompiler_builtins-b4dc7fccf0559c4c.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lc"
  = note: /usr/bin/ld: /home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libstd-60e6d7e96d924282.rlib(std-60e6d7e96d924282.std.a8eo65f6-cgu.0.rcgu.o): in function `std::sys::unix::mutex::ReentrantMutex::try_lock':
          /rustc/c5a96fb7973649807a7943e7395456db158dcab6//library/std/src/sys/unix/mutex.rs:121: undefined reference to `pthread_mutex_trylock'
          /usr/bin/ld: /home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libstd-60e6d7e96d924282.rlib(std-60e6d7e96d924282.std.a8eo65f6-cgu.0.rcgu.o): in function `<std::sys::unix::mutex::PthreadMutexAttr as core::ops::drop::Drop>::drop':
          /rustc/c5a96fb7973649807a7943e7395456db158dcab6//library/std/src/sys/unix/mutex.rs:140: undefined reference to `pthread_mutexattr_destroy'
          /usr/bin/ld: /home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libstd-60e6d7e96d924282.rlib(std-60e6d7e96d924282.std.a8eo65f6-cgu.0.rcgu.o): in function `std::sys::unix::thread_local_key::set':
          /rustc/c5a96fb7973649807a7943e7395456db158dcab6//library/std/src/sys/unix/thread_local_key.rs:16: undefined reference to `pthread_setspecific'
          /usr/bin/ld: /home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libstd-60e6d7e96d924282.rlib(std-60e6d7e96d924282.std.a8eo65f6-cgu.0.rcgu.o): in function `std::sys::unix::thread_local_key::get':
          /rustc/c5a96fb7973649807a7943e7395456db158dcab6//library/std/src/sys/unix/thread_local_key.rs:22: undefined reference to `pthread_getspecific'
          /usr/bin/ld: /home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libstd-60e6d7e96d924282.rlib(std-60e6d7e96d924282.std.a8eo65f6-cgu.0.rcgu.o): in function `std::sys::unix::thread_local_key::create':
          /rustc/c5a96fb7973649807a7943e7395456db158dcab6//library/std/src/sys/unix/thread_local_key.rs:10: undefined reference to `pthread_key_create'
          /usr/bin/ld: /rustc/c5a96fb7973649807a7943e7395456db158dcab6//library/std/src/sys/unix/thread_local_key.rs:10: undefined reference to `pthread_key_create'
          /usr/bin/ld: /home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libstd-60e6d7e96d924282.rlib(std-60e6d7e96d924282.std.a8eo65f6-cgu.0.rcgu.o): in function `std::sys::unix::thread_local_key::destroy':
          /rustc/c5a96fb7973649807a7943e7395456db158dcab6//library/std/src/sys/unix/thread_local_key.rs:27: undefined reference to `pthread_key_delete'
          /usr/bin/ld: /rustc/c5a96fb7973649807a7943e7395456db158dcab6//library/std/src/sys/unix/thread_local_key.rs:27: undefined reference to `pthread_key_delete'
          /usr/bin/ld: /home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libstd-60e6d7e96d924282.rlib(std-60e6d7e96d924282.std.a8eo65f6-cgu.0.rcgu.o): in function `std::sys::unix::rwlock::RWLock::read':
          /rustc/c5a96fb7973649807a7943e7395456db158dcab6//library/std/src/sys/unix/rwlock.rs:23: undefined reference to `pthread_rwlock_rdlock'
          /usr/bin/ld: /home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libstd-60e6d7e96d924282.rlib(std-60e6d7e96d924282.std.a8eo65f6-cgu.0.rcgu.o): in function `std::sys::unix::rwlock::RWLock::raw_unlock':
          /rustc/c5a96fb7973649807a7943e7395456db158dcab6//library/std/src/sys/unix/rwlock.rs:112: undefined reference to `pthread_rwlock_unlock'
          /usr/bin/ld: /rustc/c5a96fb7973649807a7943e7395456db158dcab6//library/std/src/sys/unix/rwlock.rs:112: undefined reference to `pthread_rwlock_unlock'
          /usr/bin/ld: /home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libstd-60e6d7e96d924282.rlib(std-60e6d7e96d924282.std.a8eo65f6-cgu.0.rcgu.o): in function `std::sys::unix::mutex::ReentrantMutex::init':
          /rustc/c5a96fb7973649807a7943e7395456db158dcab6//library/std/src/sys/unix/mutex.rs:107: undefined reference to `pthread_mutexattr_init'
          /usr/bin/ld: /rustc/c5a96fb7973649807a7943e7395456db158dcab6//library/std/src/sys/unix/mutex.rs:109: undefined reference to `pthread_mutexattr_settype'
          /usr/bin/ld: /home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libstd-60e6d7e96d924282.rlib(std-60e6d7e96d924282.std.a8eo65f6-cgu.0.rcgu.o): in function `<std::sys::unix::mutex::PthreadMutexAttr as core::ops::drop::Drop>::drop':
          /rustc/c5a96fb7973649807a7943e7395456db158dcab6//library/std/src/sys/unix/mutex.rs:140: undefined reference to `pthread_mutexattr_destroy'
          /usr/bin/ld: /home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libstd-60e6d7e96d924282.rlib(std-60e6d7e96d924282.std.a8eo65f6-cgu.0.rcgu.o): in function `std::sys::unix::thread_local_key::get':
          /rustc/c5a96fb7973649807a7943e7395456db158dcab6//library/std/src/sys/unix/thread_local_key.rs:22: undefined reference to `pthread_getspecific'
          /usr/bin/ld: /rustc/c5a96fb7973649807a7943e7395456db158dcab6//library/std/src/sys/unix/thread_local_key.rs:22: undefined reference to `pthread_getspecific'
          /usr/bin/ld: /rustc/c5a96fb7973649807a7943e7395456db158dcab6//library/std/src/sys/unix/thread_local_key.rs:22: undefined reference to `pthread_getspecific'
          /usr/bin/ld: /home/grayshade/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-musl/lib/libstd-60e6d7e96d924282.rlib(std-60e6d7e96d924282.std.a8eo65f6-cgu.0.rcgu.o): in function `std::sys::unix::thread_local_key::set':
          /rustc/c5a96fb7973649807a7943e7395456db158dcab6//library/std/src/sys/unix/thread_local_key.rs:16: undefined reference to `pthread_setspecific'
          collect2: error: ld returned 1 exit status
```

--- a/.github/workflows/release.yaml
+++ b/.github/workflows/release.yaml
@@ -121,6 +121,34 @@ jobs:
         name: dist-x86_64-unknown-linux-gnu
         path: ./dist
 
+  dist-x86_64-unknown-linux-musl:
+    name: dist (x86_64-unknown-linux-musl)
+    runs-on: ubuntu-20.04
+    env:
+      RA_TARGET: x86_64-unknown-linux-musl
+      # For some reason `-crt-static` is not working for clang without lld
+      RUSTFLAGS: "-C link-arg=-fuse-ld=lld -C target-feature=-crt-static"
+    container:
+      image: rust:alpine
+      volumes:
+      - /usr/local/cargo/registry
+
+    steps:
+    - name: Install dependencies
+      run: apk add --no-cache git clang lld musl-dev
+
+    - name: Checkout repository
+      uses: actions/checkout@v2
+
+    - name: Dist
+      run: cargo xtask dist
+
+    - name: Upload artifacts
+      uses: actions/upload-artifact@v1
+      with:
+        name: dist-x86_64-unknown-linux-musl
+        path: ./dist
+
   dist-aarch64-unknown-linux-gnu:
     name: dist (aarch64-unknown-linux-gnu)
     runs-on: ubuntu-16.04
@@ -216,7 +244,7 @@ jobs:
   publish:
     name: publish
     runs-on: ubuntu-16.04
-    needs: ['dist-x86_64-pc-windows-msvc', 'dist-aarch64-pc-windows-msvc', 'dist-x86_64-unknown-linux-gnu', 'dist-aarch64-unknown-linux-gnu', 'dist-x86_64-apple-darwin', 'dist-aarch64-apple-darwin']
+    needs: ['dist-x86_64-pc-windows-msvc', 'dist-aarch64-pc-windows-msvc', 'dist-x86_64-unknown-linux-gnu', 'dist-x86_64-unknown-linux-musl', 'dist-aarch64-unknown-linux-gnu', 'dist-x86_64-apple-darwin', 'dist-aarch64-apple-darwin']
     steps:
     - name: Install Nodejs
       uses: actions/setup-node@v1
@@ -246,6 +274,10 @@ jobs:
     - uses: actions/download-artifact@v1
       with:
         name: dist-x86_64-unknown-linux-gnu
+        path: dist
+    - uses: actions/download-artifact@v1
+      with:
+        name: dist-x86_64-unknown-linux-musl
         path: dist
     - uses: actions/download-artifact@v1
       with:

--- a/xtask/src/dist.rs
+++ b/xtask/src/dist.rs
@@ -59,7 +59,7 @@ fn dist_client(version: &str, release_tag: &str) -> Result<()> {
 
 fn dist_server() -> Result<()> {
     let target = get_target();
-    if target.contains("-linux-gnu") {
+    if target.contains("-linux-gnu") || target.contains("-linux-musl") {
         env::set_var("CC", "clang");
     }
 

